### DISCUSSION TITLE
Can applications set different lease times for different services?

### DISCUSSION THREAD

**Discussion by ndyck14** [openthread/openthread#8186] [Q&A]
There are cases where it is important to set shorter lived lease times (e.g. `_matterc._udp`) vs longer lease times (e.g. `_matter._tcp`). It appears that you can change the lease time, but its a universal time for all services rather than tracked per service. I suppose one could adjust this just before publishing, but since lease renewal is handled by OT, that seems very likely to be problematic. Is there a known way to handle this already or is this effectively a new feature request?

https://github.com/openthread/openthread/blob/ccf8ea69230bf474cbcd61ef61fc5519df7665dd/include/openthread/srp_client.h#L385

**Comment by jwhui** [comment]
As you noted, OpenThread's SRP Client currently does not really support configuring different lease times for different service instances. While the SRP protocol supports maintaining different lease times, it will require adding some additional implementation complexity to manage updating service instances that have very different lease times - it affects how we aggregate and synchronize updates to reduce the number of SRP Update interactions with the server.

@abtink , thoughts?

**Reply by abtink** [reply]
> Is there a known way to handle this already or is this effectively a new feature request?

I would say it is effectively a new feature.
What you mention (changing the lease time before adding service) can partially work (it works for first registration, but not subsequent refreshes of the same service).

---

The internal data model used by `Srp::Client` tracks the "lease renew time" per service. So may be possible to add mechanism to allow users to specify different desired lease (key-lease) intervals per service. 

I can take a look into this later. But I agree that it will, for sure, add complexity in the code (increase code size) and also some inefficiencies (e.g., on battery-powered devices - we may not be able to aggregate refreshes to reduce num of exchanged messages).

@ndyck14  is it possible for application on client side to explicitly remove (unregister) services after some time out. I feel this may be better/easier model compared to using different lease interval to expire the services on server?

**Reply by ndyck14** [reply]
interesting. Yes, if the lease renew time is tracked per service, it seems fine for the application to temporarily set the lease time to shorter just before it registers a short lived service, and set it back after, at least for the specific use-case i mentioned around `matterc._udp`.

There may still be other cases where it could be beneficial for separate lease times tracked - but if a service is to be renewed anyway and the application is expected to continue running gracefully, it is perhaps not a near term need

**Reply by abtink** [reply]
I am looking into adding general support for this into `Srp::Client` as a new feature (will work on it more when I get some time).

One thing about this is that it will be a bigger behavioral change and we need to make sure it does not break any of the functionality so make sense to do a lot of testing and validation.

**Reply by abtink** [reply]
Just adding note:
- If we have multiple services with different lease times, they need to be registered separately in different SRP update message and also refreshed in separate updates.
- The lease and key lease intervals are given in Additional Section (in an Update Lease Option) in an SRP update message and therefore are applicable to everything included in the SRP update message.

**Reply by jwhui** [reply]
> If we have multiple services with different lease times, they need to be registered separately in different SRP update message and also refreshed in separate updates.

I think this makes sense. We could try to align phases when the periods are multiples of each other (i.e. harmonics), but I'm not sure the additional complexity is worth the benefit.

**Reply by abtink** [reply]
Just submitted for this:
- https://github.com/openthread/openthread/pull/8211
